api/metrics-api-jaxrs/src/main/java/org/hawkular/metrics/api/jaxrs/handler/GaugeHandler.java

```diff
@@ -724,4 +725,164 @@ public void findPeriods(
                     .subscribe(asyncResponse::resume, t -> asyncResponse.resume(ApiUtils.serverError(t)));
         }
     }
+
+    @GET
+    @Path("/{id}/rate")
+    @ApiOperation(value = "Retrieve gauge rate data points.", response = DataPoint.class, responseContainer = "List")
+    @ApiResponses(value = {
+            @ApiResponse(code = 200, message = "Successfully fetched metric data."),
+            @ApiResponse(code = 204, message = "No metric data was found."),
+            @ApiResponse(code = 400, message = "Time range is invalid.", response = ApiError.class),
+            @ApiResponse(code = 500, message = "Unexpected error occurred while fetching metric data.",
+                    response = ApiError.class)
+    })
+    public void findRate(
+            @Suspended AsyncResponse asyncResponse,
+            @PathParam("id") String id,
+            @ApiParam(value = "Defaults to now - 8 hours") @QueryParam("start") Long start,
+            @ApiParam(value = "Defaults to now") @QueryParam("end") Long end,
+            @ApiParam(value = "Limit the number of data points returned") @QueryParam("limit") Integer limit,
+            @ApiParam(value = "Data point sort order, based on timestamp") @QueryParam("order") Order order
+    ) {
+        TimeRange timeRange = new TimeRange(start, end);
+        if (!timeRange.isValid()) {
+            asyncResponse.resume(badRequest(new ApiError(timeRange.getProblem())));
+            return;
+        }
+
+        MetricId<Double> metricId = new MetricId<>(tenantId, GAUGE, id);
+
+        if (limit == null) {
+            limit = 0;
+        }
+        if (order == null) {
+            order = Order.defaultValue(limit, start, end);
```

Comment:
Because we need to know the value of the limit parameter. Can you access
other param values in a converter ?

Besides, I'm not sure a converter is the right place to store a default
value logic. It should simply convert a string to another string. Actually,
if the param is absent, I don't think RestEasy will invoke the converter.
When possible, a @Default annotation with the string representation is
better, I think.

2016-05-10 12:02 GMT+02:00 Michael Burman <notifications@github.com>:

> In
> api/metrics-api-jaxrs/src/main/java/org/hawkular/metrics/api/jaxrs/handler/GaugeHandler.java
> <https://github.com/hawkular/hawkular-metrics/pull/501#discussion_r62643773>
> :
>
> > +            @ApiParam(value = "Limit the number of data points returned") @QueryParam("limit") Integer limit,
> > +            @ApiParam(value = "Data point sort order, based on timestamp") @QueryParam("order") Order order
> > +    ) {
> > +        TimeRange timeRange = new TimeRange(start, end);
> > +        if (!timeRange.isValid()) {
> > +            asyncResponse.resume(badRequest(new ApiError(timeRange.getProblem())));
> > +            return;
> > +        }
> > +
> > +        MetricId<Double> metricId = new MetricId<>(tenantId, GAUGE, id);
> > +
> > +        if (limit == null) {
> > +            limit = 0;
> > +        }
> > +        if (order == null) {
> > +            order = Order.defaultValue(limit, start, end);
>
> Why isn't OrderConverter forcing these default values and avoiding null?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/hawkular/hawkular-metrics/pull/501/files/8dc33a2df7c4b5e24ee658da82928ea1620b20ce#r62643773>
>



-- 
Thomas Segismont
JBoss ON Engineering Team